src/arch/arm/32/machine/capdl.c

```diff
@@ -265,8 +265,8 @@ void print_cap_arch(cap_t cap)
     switch (cap_get_capType(cap)) {
     case cap_page_table_cap: {
         asid_t asid = cap_page_table_cap_get_capPTMappedASID(cap);
-        findPDForASID_ret_t find_ret = findPDForASID(asid);
-        vptr_t vptr = cap_page_table_cap_get_capPTMappedAddress(cap);
+        UNUSED findPDForASID_ret_t find_ret = findPDForASID(asid);
```

Comment:
nitpicking: can you make adding the `UNUSED` a separate commit? This seems not directly to `CONFIG_PRINTING`, `CONFIG_DEBUG_BUILD`.

Reply:
CONFIG_DEBUG_BUILD enables this code but !CONFIG_PRINTING means that these variables become unused when printf is #ifdef'd out.

Reply:
Instead of flagging everything with UNUSED explicitly, we could also ensure that the printf() dummy-implementation for `!CONFIG_PRINTING` does this. What do you thing of the approach on https://github.com/axel-h/seL4/tree/patch-axel-1a that I'm currently playing with for better RISC-V SBI console handling

Reply:
In this case, the C preprocessor removes the entire printf statement so that it isn't seen by the proofs. Maybe something like the following would be able to be handled by automations in the proof, but it would still get trapped by the preprocessor verification tests as the kernel AST would still be different:
```
#ifdef CONFIG_PRINTING
#define printf(a, b, c) kernel_printf(a,b,c)
#else
#define printf(a, b, c) {UNUSED_VARIABLE(a); UNUSED_VARIABLE(b); UNUSED_VARIABLE(c);}
#endif
```

Reply:
So the goal really is to remove the `printf()` (and related downstream functions) from the code completely,  so the proofs don't see them at all? For the C code I consider it is much more elegant if we don't need macros at all, but let the compiler handle this completely and conclude that all this code can be removed (unless optimization is completely disabled). AM I too naive to assume the proofs conclude that this code does nothing, so there is nothing to worry about?

Reply:
> So the goal really is to remove the `printf()` (and related downstream functions) from the code completely, so the proofs don't see them at all? 

Yes

> For the C code I consider it is much more elegant if we don't need macros at all, but let the compiler handle this completely and conclude that all this code can be removed (unless optimization is completely disabled). Am I too naive to assume the proofs conclude that this code does nothing, so there is nothing to worry about?

`printf` (specifically its use of varargs) is not in the verification subset and therefore doesn't have any semantics in the toolchain. This means the verification toolchain will reject the code before it can figure out that it's unused.

You can't fully rely on the compiler to remove code that is unused either, and I'm reasonably sure that UB even in code that is unused at runtime is grounds for the compiler to do whatever it likes. (Not that this is UB, but since the verification tool chain doesn't have semantics for it, we can't prove that it isn't, so there is not really a different choice for code that is not in the verification subset).

Reply:
> Maybe something like the following would be able to be handled by automations in the proof, but it would still get
> trapped by the preprocessor verification tests as the kernel AST would still be different:

```
#ifdef CONFIG_PRINTING
#define printf(a, b, c) kernel_printf(a,b,c)
#else
#define printf(a, b, c) {UNUSED_VARIABLE(a); UNUSED_VARIABLE(b); UNUSED_VARIABLE(c);}
#endif
```

This depends on what `a`, `b`, and `c` end up as. If they are complex expressions, potentially with function calls, then that would incur verification work, because they'd still have to be checked for safety even if they are then ignored. If they are always simple, then it would likely not be a big problem.

Reply:
That makes we wonder about places in the code like
```
word_t UNUSED foo = do_something_to get_foo();
printf("foo is %p", foo)
```
where `do_something_to get_it()` could also be executed even if `printf()` is def'd out, since - as you said - the compiler will not remove it. Isn't this something that is also annoying for the proof because that code should not be there either? Seems to me the only really sane way out here is enclosing all calls to `printf()` in `#ifdef CONFIG_PRINTING` and also wrap all the variables used there. And then don't even have a macro for define `printf()` and friends away, but just dont provide it and the cost must pass without warnings.

Reply:
Yes, that is also visible to verification. It's not necessarily very annoying, because if it's just a function call we're likely to already have a lemma about that, but it's easier if it's not visible (and it can be a problem if it is complex)